Account settings: Change password

danielebogo opened the pull request:
Ref. #6561 

![changing-password](https://user-images.githubusercontent.com/912252/50837436-88cf2180-1353-11e9-806d-9bf6ca220cba.gif)

This PR adds a small enhancement to the app with a new setting option that allows the user to change the password.

To test:
- Select the tab _Me_ -> _Account Settings_
- Select _Change Password_
- Write a new password. _Save_ bar button item is enabled only if the password is at least 6 chars length

Update release notes:

- [x] I have added an item to `RELEASE-NOTES.txt`.

aerych commented:
Heya @danielebogo.  Looking good :)  I tested and was able to change my password via the feature.  I have a couple of early thoughts before I dig into the code review. 

I'm wondering about the position of the change password option in the list.  Would it be better to have it after username and email address since it tends to be that way on our login and sign up forms?  Curious for @SylvesterWilmott thoughts here.   Also, should each word be capitalized, both on the settings screen and the change password screen itself for consistency with the other options on the settings screen?
![simulator screen shot - iphone 8 - 2019-01-08 at 10 42 33](https://user-images.githubusercontent.com/1435271/50846336-84474080-1334-11e9-9a14-3f0f34c5b942.png)

I noticed that if I hit enter on my mac's keyboard (like an external keyboard) or if I tap Done on the app's keyboard that the change password view controller just dismisses without attempting to change the password.  I wonder if a user would think they'd changed their password when they really hadn't.  Could we wire up the Done button to also save the password just in case? 

Thoughts?

iamthomasbishop commented:
I largely have the same questions as @aerych, but also one more thing that I'd like to hear your thoughts on. 

Should we have:

1. A second password field as a confirmation/validation, and/or...
2. A "current password" field? 

It seems like more often than not, when I'm resetting my password on a service, I have to do one or both of those. 1) so I don't mis-type my new password and 2) for added security?

danielebogo commented:
> I'm wondering about the position of the change password option in the list. Would it be better to have it after username and email address since it tends to be that way on our login and sign up forms?

I thought the same at the beginning but then I followed the Android implementation.

> Also, should each word be capitalized, both on the settings screen and the change password screen itself for consistency with the other options on the settings screen?

Yes absolutely!

@iamthomasbishop if I'm not wrong we have one filed only also in Calypso right? 🤔
@aerych Good catch about the Done button. At this point, because the _Save_ button is enabled only if the password length is => 6, I don't know if apply the same logic to it or block it anyway. Any thoughts?

nheagy commented on WordPress/Classes/ViewRelated/Me/ChangePasswordViewController.swift:
```diff
@@ -0,0 +1,73 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        mode = .password
```
This should be `.newPassword`.

I believe to get iOS 12's password generation to work there also has to be a textfield containing the username with `mode = .username`, however that will require some experimenting.

danielebogo replied:
New `SettingsTextModes` added: `SettingsTextModesNewPassword`. This will add:
```
if (@available(iOS 12.0, *)) {
   self.textField.textContentType = UITextContentTypeNewPassword;
}
```
Thanks for the hint!

iamthomasbishop commented:
@aerych @danielebogo  Fair enough, thanks for the context!

ScoutHarris commented:
Works great for me. Nice addition @danielebogo ! I'll review again when requested changes made.

danielebogo commented:
Thanks @ScoutHarris and @aerych for reviewing this. I Fixed the text and the _Done_ button behaviour. I also moved the _Change Password_ row at the 3rd position after _Username_ and _Email_. @iamthomasbishop @SylvesterWilmott what do you think?

@nheagy thanks for the hint! I added a new mode and setup the _.newPassword_ `textContentType`. But you are right when you said I need the username field to make it work. I'm playing with it now but I don't know if do it in another branch and wait your PR first. What do you think?

danielebogo commented:
@SylvesterWilmott @iamthomasbishop 

![simulator screen shot - iphone x - 2019-01-09 at 14 29 30](https://user-images.githubusercontent.com/912252/50905580-0f026b00-141b-11e9-8973-425b384d07a3.png)

![simulator screen shot - iphone x - 2019-01-09 at 14 29 36](https://user-images.githubusercontent.com/912252/50905600-14f84c00-141b-11e9-8156-33cafaa4f558.png)

SylvesterWilmott commented:
LGTM @danielebogo . Good job.

nheagy commented:
For what it's worth, I found the issue for implementing the iOS 12 password changing API: #10031

ScoutHarris commented on WordPress/Classes/ViewRelated/Me/AccountSettingsViewController.swift:
```diff
@@ -128,6 +140,40 @@ private class AccountSettingsController: SettingsController {
+                            let errorMessage = error?.localizedDescription ?? Constants.changePasswordGenericError
+                            SVProgressHUD.showError(withStatus: errorMessage)
+                        }
+                        SVProgressHUD.dismiss(withDelay: TimeInterval(2))
```
Can we increase the time the error is displayed? 2 seconds is not really long enough, especially when a long error message is displayed, like this: 

<img width="203" alt="change_password_error" src="https://user-images.githubusercontent.com/1816888/50923920-541ea100-140b-11e9-825a-7409f481f447.png">

I believe elsewhere we've started using 5 seconds, which feels much better. But it's up to you.

danielebogo replied:
Ah ok! Good to know! I'll remove it!

danielebogo commented:
@ScoutHarris thanks for the hint! I removed that line so it should take the 5 seconds (default value)

ScoutHarris commented:
Awesome! Thanks @danielebogo ! :shipit:  from me.
